src/lib.rs

```diff
@@ -49,47 +53,151 @@ const FALLBACK_FAVICON_ICO: &[u8] = include_bytes!("../spin-favicon.ico");
 const BUFFER_SIZE: usize = 64 * 1024;
 const DEFLATE_LEVEL: flate2::Compression = flate2::Compression::fast();
 
+#[derive(PartialEq)]
+struct ContentEncoding {
+    // We limit expressed encodings to ones that we support
+    encoding: SupportedEncoding,
+    weight: Option<f32>,
+}
+
+impl fmt::Display for ContentEncoding {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.encoding)?;
+
+        if let Some(weight) = self.weight {
+            write!(f, ";q={weight}")?;
+        }
+
+        Ok(())
+    }
+}
+
+impl PartialEq<SupportedEncoding> for ContentEncoding {
+    fn eq(&self, other: &SupportedEncoding) -> bool {
+        self.encoding == *other
+    }
+}
+
+impl PartialEq<SupportedEncoding> for &ContentEncoding {
+    fn eq(&self, other: &SupportedEncoding) -> bool {
+        self.encoding == *other
+    }
+}
+
+impl PartialOrd for ContentEncoding {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        let aweight = self.weight.unwrap_or(1.0);
+        let bweight = other.weight.unwrap_or(1.0);
+        match aweight.partial_cmp(&bweight) {
+            Some(Ordering::Equal) => match (self.encoding, other.encoding) {
+                // Always prefer brotli
+                (SupportedEncoding::Brotli, _) => Some(Ordering::Greater),
+                (_, SupportedEncoding::Brotli) => Some(Ordering::Less),
+                // Otherwise prefer the more specific option
+                (SupportedEncoding::None, _) => Some(Ordering::Less),
+                (_, SupportedEncoding::None) => Some(Ordering::Greater),
+                // Everything else is roughly equal
+                (_, _) => Some(Ordering::Equal),
+            },
+            v => v,
+        }
+    }
+}
+
+impl FromStr for ContentEncoding {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split(';');
+        let v = parts.next().unwrap().trim();
```

Comment:
nit: I think we should prefer not using single character binding names in a somewhat large function like this:

```suggestion
        let encoding = parts.next().unwrap().trim();
```